hashintel/hash
# H-867, H-1008: Implement User and Org customizable areas, supporting blocks

## 🌟 What is the purpose of this PR?




This PR adds an editable bio to the profile page, supporting blocks similar to the page editor.

Drive-by changes:

- [Move "View all pages ->" into the "Page >" section](https://linear.app/hash/issue/H-1008/move-view-all-pages-into-the-page-section)

## 🔗 Related links





- H-867
- H-1008

## 🔍 What does this change?




- A new `BlockCollection` entity type has been introduced, which both the `Page` and the new `ProfileBio` entity types inherit from. `BlockCollection` defines the outgoing `contains` link which links it to blocks.
- The existing `updatePageContents` GQL resolver has been replaced with `updateBlockCollectionContents`, decoupling its functionality from pages.
- The existing `save.ts` script has been modified to decouple its functionality from pages.
- The existing `PageBlock` component has been replaced by a `BlockCollection` component, which removes a variety of page-specific functionality to make it possible to render a collection of blocks outside of the page editor

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?




This PR:


### 📜 Does this require a change to the docs?




The changes in this PR:


### 🕸️ Does this require a change to the Turbo Graph?




The changes in this PR:


## ⚠️ Known issues



- When switching between editing/viewing state in the bio editor, the content briefly disappears. I'm not sure exactly why this is happening, but probably something to do with how prosemirror re-mounts itself when `readonly` is toggled.
- Because saving happens automatically, and there isn't currently a convenient way of triggering it in the parent component of `BlockCollection`, it's possible that if the user presses the "check" icon before the save has been persisted that the latest changes won't be reflected in the UI. The may still have saved, but the `fetchProfileBioContents` function may have been called before the save persisted.

## 🐾 Next steps



- [Store index on block collection contents link in properties instead of link metadata](https://linear.app/hash/issue/H-1030/store-index-on-block-collection-contents-link-in-properties-instead-of)

## 🛡 What tests cover this?



None

## ❓ How to test this?



⚠️ This PR requires backend changes to be applied to `main`, so using the Vercel deployment risks breaking the production environment 

1. Go to the profile page of the logged in user
2. Click the edit icon on the bio to edit the bio. Try inserting blocks, using `@` mentions, etc
3. Go to a page, and ensure the page editor looks and functions exactly like on `main`

## 📹 Demo



https://github.com/hashintel/hash/assets/42802102/cd80b6d1-e2be-44ae-90fc-95506ae8c61e
